draft-irtf-cfrg-aead-limits.md

```diff
@@ -678,45 +706,209 @@ IA <= AEA
+-->
+
+It implies the following simplified limit, which for most reasonable values
+of `p` is dominated by a technical limitation around `q = 2^100`:
 
 ~~~
-v <= ((p / u) * 2^106) / 8l
-  <= (p * 2^103) / (l * u)
+q <= min( p * 2^246 - o,  sqrt(p) * 2^256 - B, 2^100 )
 ~~~
 
-### AEAD_AES_128_CCM and AEAD_AES_128_CCM_8
 
-The integrity limit for AEAD_AES_128_CCM is as follows.
+### Integrity Limit
+
+The AE limit for AEAD_CHACHA20_POLY1305 essentially is the integrity (multi-key)
+bound. The former hence also applies to the latter:
 
 ~~~
-v + q <= (p / u)^(1/2) * 2^63 / l
+IA <= AEA
 ~~~
 
-Likewise, the integrity limit for AEAD_AES_128_CCM_8 is as follows.
+{{mu-ccp-ae}} therefore contains the integrity limits.
+
+
+## AEAD_AES_128_CCM and AEAD_AES_128_CCM_8
+
+There are currently no concrete multi-key bounds for AEAD_AES_128_CCM or
+AEAD_AES_128_CCM_8. Thus, to account for the additional
+factor `u`, i.e., the number of keys, each `p` term in the confidentiality and
+integrity limits is replaced with `p / u`.
+
+The multi-key integrity limit for AEAD_AES_128_CCM is as follows.
+
+~~~
+v + q <= sqrt(p / u) * 2^63 / l
+~~~
+
+Likewise, the multi-key integrity limit for AEAD_AES_128_CCM_8 is as follows.
 
 ~~~
 v * 2^64 + (2l * (v + q))^2 <= (p / u) * 2^128
 ~~~
 
+
+## Multi-Key Examples
+
+An example protocol might choose to aim for a multi-key AEA, CA, and IA that is at
+most 2<sup>-50</sup>.  If the messages exchanged in the protocol are at most a
+common Internet MTU of around 1500 bytes, then a value for l might be set to
+2<sup>7</sup>.  {{ex-table-mu}} shows limits for q and v across all keys that
+might be chosen under these conditions.
+
+| AEAD                   | Maximum q                | Maximum v              |
+|:-----------------------|-------------------------:|-----------------------:|
+| AEAD_AES_128_GCM       | 2<sup>69</sup>/B         | 2<sup>69</sup>/B       |
+| AEAD_AES_256_GCM       | 2<sup>69</sup>/B         | 2<sup>69</sup>/B       |
+| AEAD_CHACHA20_POLY1305 | 2<sup>100</sup>          | 2<sup>46</sup>         |
+| AEAD_AES_128_CCM       | 2<sup>30</sup>/sqrt(u)   | 2<sup>30</sup>/sqrt(u) |
+| AEAD_AES_128_CCM_8     | 2<sup>30.9</sup>/sqrt(u) | 2<sup>13</sup>/u |
+{: #ex-table-mu title="Example multi-key limits"}
+
+The limits for AEAD_AES_128_GCM, AEAD_AES_256_GCM, AEAD_AES_128_CCM, and
+AEAD_AES_128_CCM_8 assume equal proportions for q and v. The limits for
+AEAD_AES_128_GCM, AEAD_AES_256_GCM and AEAD_CHACHA20_POLY1305 assume nonce
+randomization being deployed, like in TLS 1.3 {{TLS}} and QUIC {{?RFC9001}}.
+
+The limits for AEAD_AES_128_GCM and AEAD_AES_256_GCM further depend on the
+maximum number B of 128-bit blocks encrypted by any single key. For example,
+limiting each key to encrypt at most 2^20 (about a million) messages (of size
+around 1500 bytes as above), this results in B = 2^27 and hence limits for
+q and v of 2^42 messages that can be protected resp. forgery attempts that can
+be tolerated.
```

Comment:
Maybe instead

```suggestion
maximum number B of 128-bit blocks encrypted by any single key. For example,
limiting the number of messages to at most 2^20 (about a million) results in
B = 2^27, which limits both q and v to 2^42 messages.
```

Reply:
The 1500 bytes message size <= 2^7 blocks entered my calculation as B = 2^20 * 2^7, so I guess that should still be mentioned?

```suggestion
maximum number B of 128-bit blocks encrypted by any single key. For example,
limiting the number of messages (of size <= 2^7 blocks) to at most 2^20 (about a
million) per key results in B = 2^27, which limits both q and v to 2^42 messages.
```